packages/cpp/src/browser/cpp-build-configurations.ts

```diff
@@ -0,0 +1,172 @@
+import { StorageService } from "@theia/core/lib/browser/storage-service";
+import { CppPreferences } from "./cpp-preferences";
+
+export interface CppBuildConfiguration {
+    name: string;
+    directory: string;
+}
+
+/** What we save in the local storage.  */
+class SavedActiveBuildConfiguration {
+    configName?: string;
+}
+
+@injectable()
+export class CppBuildConfigurations {
+    @inject(StorageService)
+    protected readonly storageService: StorageService;
+
+    @inject(CppPreferences)
+    protected readonly cppPreferences: CppPreferences;
+
+    /**
+     * The current active build configuration.  undefined means there's not
+     * current active configuration.
+     */
+    protected activeConfig: CppBuildConfiguration | undefined;
+
+    /** Emitter for when the active build configuration changes.  */
+    protected readonly activeConfigChangeEmitter = new Emitter<CppBuildConfiguration | undefined>();
+
+    readonly ACTIVE_BUILD_CONFIGURATION_STORAGE_KEY = 'cpp.active-build-configuration';
+    readonly BUILD_CONFIGURATIONS_PREFERENCE_KEY = 'cpp.buildConfigurations';
+
+    public ready: Promise<void>;
+
+    @postConstruct()
+    init() {
+        // Try to read the active build config from local storage.
+        this.ready = new Promise(resolve => {
+            this.loadActiveConfiguration().then(resolve);
+        });
+
+        // However, preferences may not have been loaded yet, so register a
+        // listener for when they are, and try loading the active config.
+        this.cppPreferences.onPreferenceChanged(e => {
+            if (e.preferenceName === this.BUILD_CONFIGURATIONS_PREFERENCE_KEY) {
+                this.loadActiveConfiguration();
+            }
+        });
+
+    }
+
+    /** Load the active build config from the persistent storage.  */
+    protected async loadActiveConfiguration() {
+        const savedConfig =
+            await this.storageService.getData<SavedActiveBuildConfiguration>(
+                this.ACTIVE_BUILD_CONFIGURATION_STORAGE_KEY);
+
+        if (savedConfig !== undefined && savedConfig.configName !== undefined) {
+            // Try to find an existing config with that name.
+            const configs = this.getConfigs();
+            const config = configs.find(cfg => savedConfig.configName === cfg.name);
+            if (config) {
+                this.setActiveConfig(config);
+            }
+        }
+    }
+
+    /**
+     * Save the active build config name to persistent storage.
+     */
+    protected saveActiveConfiguration(config: CppBuildConfiguration | undefined) {
+        this.storageService.setData<SavedActiveBuildConfiguration>(
+            this.ACTIVE_BUILD_CONFIGURATION_STORAGE_KEY, {
+                configName: config ? config.name : undefined,
+            });
+    }
+
+    /** Get the active build configuration.  */
+    getActiveConfig(): CppBuildConfiguration | undefined {
+        return this.activeConfig;
```

Comment:
In the local storage, we only store the name.  So to implement getActiveConfig this way, we would have to

1. Get the name of the active config from the local storage
2. Search the config structure by name from the preferences

That sounds like a lot of work for something that we can just store in a field, but in itself I don't think it would be a problem.

A more important problem would depend on how we handle build configurations that disappear.  Let's say I define a build config named "foo", make it active, then remove it from my preferences.  If we want to make it so "foo" stays active as long as you don't choose another config the be the active one, we have to store it somewhere.  Otherwise, when we'll look up "foo" in the build config list in the preferences, we won't find anything.

Reply:
Makes sense, thank you.